packages/styled-components/src/sheet/test/WindowedTag.test.js

```diff
@@ -0,0 +1,192 @@
+
+  it('Correctly builds the linked list', () => {
+    const virtualTag = new VirtualTag();
+    const windowedTag1 = new WindowedTag(virtualTag);
+
+    expect(windowedTag1.prev).toBe(undefined);
+    expect(virtualTag.window).toBe(windowedTag1);
+    const windowedTag2 = new WindowedTag(virtualTag);
+    expect(windowedTag2.prev).toBe(windowedTag1);
+    expect(virtualTag.window).toBe(windowedTag2);
+  });
+
+  it('Inserts rules one window at a time', () => {
+    const virtualTag = new VirtualTag();
+    const windowedTag1 = new WindowedTag(virtualTag);
+    const windowedTag2 = new WindowedTag(virtualTag);
+
+    // Insertions into the windowedTag1
+    expect(windowedTag1.insertRule(0, '.b {}')).toBe(true);
+    expect(windowedTag1.insertRule(0, '.a {}')).toBe(true);
+    expect(windowedTag1.insertRule(2, '.c {}')).toBe(true);
+    expect(windowedTag1.getRule(0)).toBe('.a {}');
+    expect(windowedTag1.getRule(1)).toBe('.b {}');
+    expect(windowedTag1.getRule(2)).toBe('.c {}');
+    expect(windowedTag1.getRule(3)).toBe('');
+    expect(windowedTag1.length).toBe(3);
+
+    // Check some rules in the tag
+    expect(virtualTag.getRule(2)).toBe('.c {}');
+    expect(virtualTag.length).toBe(3);
+
+    // Insertions into the windowedTag2
+    expect(windowedTag2.insertRule(0, '.b-second {}')).toBe(true);
+    expect(windowedTag2.insertRule(0, '.a-second {}')).toBe(true);
+    expect(windowedTag2.insertRule(2, '.c-second {}')).toBe(true);
+    expect(windowedTag2.getRule(0)).toBe('.a-second {}');
+    expect(windowedTag2.getRule(1)).toBe('.b-second {}');
+    expect(windowedTag2.getRule(2)).toBe('.c-second {}');
+    expect(windowedTag2.getRule(3)).toBe('');
+    expect(windowedTag2.length).toBe(3);
+
+    // Check some rules in the tag
+    expect(virtualTag.getRule(2)).toBe('.c {}');
+    expect(virtualTag.getRule(5)).toBe('.c-second {}');
+    expect(virtualTag.length).toBe(6);
+  });
+
+  it('Can interleave rule insertions between two windows', () => {
+    const virtualTag = new VirtualTag();
+    const windowedTag1 = new WindowedTag(virtualTag);
+    const windowedTag2 = new WindowedTag(virtualTag);
+
+    expect(windowedTag1.insertRule(0, '.b {}')).toBe(true);
+    expect(windowedTag2.insertRule(0, '.b-second {}')).toBe(true);
+    expect(windowedTag1.insertRule(0, '.a {}')).toBe(true);
+
+    // Check some rules in the tag
+    expect(virtualTag.getRule(2)).toBe('.b-second {}');
+    expect(virtualTag.length).toBe(3);
+
+    expect(windowedTag2.insertRule(0, '.a-second {}')).toBe(true);
+    expect(windowedTag1.insertRule(2, '.c {}')).toBe(true);
+    expect(windowedTag2.insertRule(2, '.c-second {}')).toBe(true);
+
+    expect(windowedTag1.getRule(0)).toBe('.a {}');
+    expect(windowedTag1.getRule(1)).toBe('.b {}');
+    expect(windowedTag1.getRule(2)).toBe('.c {}');
+    expect(windowedTag1.length).toBe(3);
+
+    // The current implementation allows windows to "bleed" if you request an offset that lands in another window
```

Comment:
@kitten - I noticed while testing that the "windows" can "bleed" into other windows, but I don't think that's actually a bad thing? We could guard against it, but I don't know that it will be something we need to worry about in actual usage?

Reply:
Yea, I figured given the current constraints this shouldn’t ever happen anyway 👍